src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java

```diff
@@ -379,8 +380,18 @@ protected void moveCaret(MouseEvent e) {
     public void focusGained(FocusEvent e) {
         if (component.isEnabled()) {
             if (component.isEditable()) {
-                setVisible(true);
+                if (savedBlinkRate != 0) {
+                    setBlinkRate(savedBlinkRate);
+                    savedBlinkRate = 0;
+                }
+            } else {
+                if (getBlinkRate() != 0) {
+                    savedBlinkRate = getBlinkRate();
```

Comment:
Look like this can override blink rate set by user:

```java
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JTextField;
import javax.swing.SwingUtilities;
import java.awt.BorderLayout;
import java.awt.Component;
import java.awt.Robot;
import java.lang.reflect.InvocationTargetException;

public class CaretThing {

    static JFrame frame;
    static JTextField textField;
    static Robot robot;
    static JButton button;

    static void printBlinkRate() {
        System.err.println("Caret blink rate: " + textField.getCaret().getBlinkRate());
    }

    static void requestFocus(Component component) throws InterruptedException, InvocationTargetException {
        robot.waitForIdle();
        robot.delay(300);

        SwingUtilities.invokeAndWait(() -> {
            System.err.println("Requesting focus on " + component);
            printBlinkRate();
            component.requestFocus();
            printBlinkRate();
        });
    }

    static void changeEditable() throws InterruptedException, InvocationTargetException {
        robot.waitForIdle();
        robot.delay(300);
        SwingUtilities.invokeAndWait(()-> {
            System.err.println("Changing editable");
            printBlinkRate();
            textField.setEditable(!textField.isEditable());
            printBlinkRate();
        });
    }

    public static void main(String[] args) throws Exception {
        robot = new Robot();
        robot.setAutoWaitForIdle(true);
        robot.setAutoDelay(50);

        SwingUtilities.invokeAndWait(() -> {
            frame = new JFrame("Hey");
            frame.setLocationRelativeTo(null);
            textField = new JTextField("Some long field value");

            textField.setEditable(false);
            frame.setLayout(new BorderLayout());
            frame.add(textField, BorderLayout.NORTH);

            button = new JButton("Button");
            button.addActionListener((e)-> printBlinkRate());

            frame.add(button, BorderLayout.CENTER);
            frame.pack();
            frame.setVisible(true);

            button.requestFocus();
            textField.getCaret().setBlinkRate(2500);
        });

        requestFocus(textField);

        requestFocus(button); // comment this to get 0 blink rate on editable field

        changeEditable();

        requestFocus(textField);

        SwingUtilities.invokeAndWait(CaretThing::printBlinkRate);
    }
}
```
Running the sample above at least two possible misbehaviors(tested on Linux):
* you will receive blink rate `500` instead of user set `2500`
* but if you comment `requestFocus(button);` line then blink rate will be `0` (it shows non-blinking cursor)

<hr>

Other than that I am worried about `getBlinkRate()` does not return the same value as passed to`setBlinkRate()` (e.g. case when component is not editable, possible JCK issue?)

Reply:
I updated the code now the behaviour is correct (as far as i tested) and the blink rate is reported as if it is set to the correct value even on non-editable text element.

Reply:
I am still having issue with non-blinking cursor on editable field with the latest version of the fix.

Fails for me with test provided above and commented `requestFocus(button);` line.

Reply:
The latest version does not work good for me,
e.g. with the test above and `textField.getCaret().setBlinkRate(250)`(not `2500`):

1. with `requestFocus(button);`:
 caret blinks several times and stops
 on mouse move, focus gained on text field the caret blinks several times and stops again.
2. without `requestFocus(button);`:
 it doesn't blink until focus out / focus in on the text component
 after this blinks several times and stops.

Reply:
That's weird, does not happen on any of my systems. Which OS are you testing on?

Reply:
Ok, i tries on Ubuntu 22 on x11 desktop and i see this behavior. However - i checked out master and i also can see the same glitch so this is not related to my fix.

Reply:
Can confirm, filed [JDK-8299050](https://bugs.openjdk.org/browse/JDK-8299050) and [JDK-8299051](https://bugs.openjdk.org/browse/JDK-8299051) to track them.